homeassistant/components/homekit/thermostats.py

```diff
@@ -0,0 +1,214 @@
+    ATTR_ENTITY_ID, ATTR_UNIT_OF_MEASUREMENT,
+    ATTR_SUPPORTED_FEATURES, TEMP_CELSIUS, TEMP_FAHRENHEIT)
+from homeassistant.helpers.event import async_track_state_change
+
+from . import TYPES
+from .accessories import HomeAccessory, add_preload_service
+from .const import (
+    SERV_THERMOSTAT, CHAR_CURRENT_HEATING_COOLING,
+    CHAR_TARGET_HEATING_COOLING, CHAR_CURRENT_TEMPERATURE,
+    CHAR_TARGET_TEMPERATURE, CHAR_TEMP_DISPLAY_UNITS,
+    CHAR_COOLING_THRESHOLD_TEMPERATURE, CHAR_HEATING_THRESHOLD_TEMPERATURE)
+
+_LOGGER = logging.getLogger(__name__)
+
+UNIT_HASS_TO_HOMEKIT = {TEMP_CELSIUS: 0, TEMP_FAHRENHEIT: 1}
+UNIT_HOMEKIT_TO_HASS = {c: s for s, c in UNIT_HASS_TO_HOMEKIT.items()}
+THC_HASS_TO_HOMEKIT = {'off': 0, STATE_HEAT: 1, STATE_COOL: 2, STATE_AUTO: 3}
+THC_HOMEKIT_TO_HASS = {c: s for s, c in THC_HASS_TO_HOMEKIT.items()}
+CHC_HASS_TO_HOMEKIT = {'off': 0, STATE_HEAT: 1, STATE_COOL: 2}
+CHC_HOMEKIT_TO_HASS = {c: s for s, c in CHC_HASS_TO_HOMEKIT.items()}
+
+
+@TYPES.register('Thermostat')
+class Thermostat(HomeAccessory):
+    """Generate a Thermostat accessory for an alarm control panel."""
+
+    def __init__(self, hass, entity_id, display_name):
+        """Initialize a Thermostat accessory object."""
+        super().__init__(display_name, entity_id, 'THERMOSTAT')
+
+        self._hass = hass
+        self._entity_id = entity_id
+        self._call_timer = None
+
+        self.heat_cool_flag_target_state = False
+        self.temperature_flag_target_state = False
+        self.cooling_thresh_temp_flag_target_state = False
+        self.heating_thresh_temp_flag_target_state = False
+
+        self.service_thermostat = add_preload_service(self, SERV_THERMOSTAT)
+        self.char_current_heat_cool = self.service_thermostat. \
+            get_characteristic(CHAR_CURRENT_HEATING_COOLING)
+        self.char_current_heat_cool.value = 0
+        self.char_target_heat_cool = self.service_thermostat. \
+            get_characteristic(CHAR_TARGET_HEATING_COOLING)
+        self.char_target_heat_cool.value = 0
+        self.char_target_heat_cool.setter_callback = self.set_heat_cool
+
+        self.char_current_temp = self.service_thermostat. \
+            get_characteristic(CHAR_CURRENT_TEMPERATURE)
+        self.char_current_temp.value = 0.0
+        self.char_target_temp = self.service_thermostat. \
+            get_characteristic(CHAR_TARGET_TEMPERATURE)
+        self.char_target_temp.value = 0.0
+        self.char_target_temp.setter_callback = self.set_target_temperature
+
+        # If the device supports it: add high and low temperature
+        self.char_cooling_thresh_temp = None
+        self.char_heating_thresh_temp = None
+        state = self._hass.states.get(self._entity_id)
+        if state is not None:
+            supported_features = state.attributes[ATTR_SUPPORTED_FEATURES]
+            if supported_features is not None and \
+                    supported_features & SUPPORT_TARGET_TEMPERATURE_HIGH:
```

Comment:
This should be done in `__init__.get_accessory` and it should be a new type (one that supports thresholds).

Reply:
Ok, I will do the feature support check in `__init__.get_accessory`. What do you mean with a new type? Can I just pass a constructor parameter to Thermostat saying if it supports this feature or not? Or should I make a "new" type that inherits from Thermostat? Or make a completely new type for it (causing a lot of code duplication obviously)? I will make all requested changes tonight (Belgian time)

Reply:
I haven't figured out the best way jet, but the issue is that as soon as you add additional characteristics to `SERV_THERMOSTAT`, HomeKit assumes that the device supports those.

I'm not sure if one of those ways would be better. The reason I suggested to test it in `get_accessory` is that this is the option I have chosen for `sensors`. Would work as well if it would be before `add_preload_service`. Choose the one that you find most convenient .